simpegEM1D/Survey.py

```diff
@@ -421,17 +421,15 @@ def projectFields(self, u):
         if self.wave_type == 'stepoff':
             # Compute EM responses
             if u.size == self.n_frequency:
-                resp = np.empty(self.n_time, dtype=float)
                 resp, _ = ffht(
                     u.flatten()*factor, self.time,
                     self.frequency, self.ftarg
                 )
             # Compute EM sensitivities
             else:
                 resp = np.zeros(
-                    (self.n_time, self.n_layer), dtype=float, order='F'
-                )
-                resp_i = np.empty(self.n_time, dtype=float)
```

Comment:
I see. So, it is unecessary to form empty array.

Reply:
It depends.
lets say I have a function that returns A

```python
def func():
    X = np.random.randn(10,10)
    return X
```

If I call func, then a new array is created X and a variable y is assigned to that memory after the function finishes.

```python
y = func()
```

If I first create an empty array (or zeros) and then call the function, there are two ways I could do it

```python
y = np.zeros([10, 10])  # or np.empty([10, 10])
y = func()
```

this would create memory for y, enter the function, create NEW memory X, and then reassign the variable(or label) y to that newly created memory, and the first ```np.zeros``` call is garbage collected. So there was no point creating it in the first place.

If instead I do

```python
y = np.zeros([10, 10])
y[:, :] = func()
```

y is created, enter function, create NEW memory X, but then copy those values into the space of the existing variable y.  The memory of X that was create inside the func is now garbage collected instead.

By default, numpy arrays are ordered with order='C' not order='F', and we want order='F' because we are using a Fortran backend. We could still use order='C' but when we pass those arrays through to Fortran there is a hidden copy of memory that occurs in order to map it to the correct ordering, something we don't want to happen every time we call the Fortran codes.

So, rather than me editing all the functions that return an array and adding order='F', I just edited the instantiation of that memory, and copy the results of functions into those placeholders.  Since we re-use those arrays, doing this once at the beginning does not incur any real cost to the code. Hence why I keep np.empty only when we have order='F' in the memory instantiation.

Clear as mud?

Reply:
I see not it's crystal clear. 

For below case, 
```python
y = np.zeros([10, 10])
y[:, :] = func()
```
when copying output from `func()` to `y`, are we using doubled amount of memory?

Reply:
I don't think it uses more memory, but the pre-allocation costs you unnecessary time.
![2018-10-22_02](https://user-images.githubusercontent.com/8020943/47303902-cf31a180-d5ea-11e8-87b4-e0149d5cd3b4.jpg)

Reply:
Here the first two cells. I cannot attach .py or .ipynb, apparently.

![2018-10-22_03](https://user-images.githubusercontent.com/8020943/47304059-2c2d5780-d5eb-11e8-95a8-8489edb4e17f.jpg)

Reply:
Interested though if @leonfoks agrees or not...

Reply:
(The pre-allocation is not costing you lots of time, it is the indexing later by filling in the new values into the existing array. If you just create it and then overwrite it you don't loose much, but it is still unnecessary. See http://nbviewer.jupyter.org/gist/prisae/af1b40ef4a2e16e130b39f6d7957be50 for an updated version.)

Reply:
got it. it seems now pretty clear to me what's the big kid on the block!